### PR TITLE
Create RandomBase class and FakeRandom and StdRandom classes inherit from it

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

Create the `RandomBase` class and update `FakeRandom` and `StdRandom` classes to inherit from it. Start using base class in code.

This is part of #3700.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

laptop, Ubuntu 20.04

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes/No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/Utilities/RandomGenerator.h [hunk 1]
```diff
@@ -38,29 +38,18 @@ uint32_t make_seed(int i, int n);
...
-using RandomGenerator = FakeRandom;
-extern RNGThreadSafe<RandomGenerator> fake_random_global;
+using RandomGenerator = FakeRandom<OHMMS_PRECISION_FULL>;
+extern RandomGenerator fake_random_global;
 #define Random fake_random_global
 #else
 // real RNG redirection
 using RandomGenerator = StdRandom<OHMMS_PRECISION_FULL>;
-extern RNGThreadSafe<RandomGenerator> random_global;
+extern RandomGenerator random_global;
```
Comment (markdewing) [inline src/Utilities/RandomGenerator.h:52]: Is it safe to drop the RNGThreadSafe wrapper?

Comment (quantumsteve) [reply src/Utilities/RandomGenerator.h:52]: I moved the wrapper into `StdRandom::operator()()` and `FakeRandom::operator()()`

Comment (ye-luo) [reply src/Utilities/RandomGenerator.h:52]: In the old code, critical is only needed on the random_global. Now the cost is paid on very call to generate a RN.

Comment (markdewing) [reply src/Utilities/RandomGenerator.h:52]: Then is it safe to have that critical section used everywhere?  It looks like previously the wrapper is only used to lock the global RNG instance.

Comment (quantumsteve) [reply src/Utilities/RandomGenerator.h:52]: Yep, that change looks like a mistake. I'll revert it.

Comment (quantumsteve) [reply src/Utilities/RandomGenerator.h]: The `RNGThreadSafe` wrapper should work as before.

**Comment by markdewing** [comment]
Test this please

**Comment by markdewing** [comment]
Test this please

**Comment by quantumsteve** [comment]
Test this please

**Review by PDoakORNL — COMMENTED** [review]
I'm a little concerned about functions such as NonLocalECPotential::mw_evaluateImpl where the single rng() calls really really stack up.  Unless the compiler does something clever this looks like a very large number of calls that now have the overhead of being virtual.

This is not the same as our higher level virtual functions that say get called once per walker per step. Rng being virtual creates is by my estimation 2-5 orders of magnitude more vtable hits per walker per move.

Changing the RandomBase API a bit for makeGaussRandomWithEngine can mitigate the virtual overhead there by getting blocks of random number to use in assigns without hitting the vtable every single call.

I have a comment on bare pointers used where references would be better style in AFQMC.

**Comment by PDoakORNL** [comment]
To expand on the previous I think whether this is a performance impacting change or not boils down to whether in the hot loops, for a particular architecture the indirect branch prediction successfully elides the vtable lookup or not. If it does then the impact will be basically 0 since the address can be predicted with 100% accuracy after the first call i.e. we never have a hot loop that uses both rng() implementations. One would hope that for any modern silicon with an ok compiler this would be true. But arm, power?

**Comment by quantumsteve** [comment]
> I'm a little concerned about functions such as NonLocalECPotential::mw_evaluateImpl where the single rng() calls really really stack up. Unless the compiler or hardware does something clever this looks like a very large number of calls that now have the overhead of being virtual.
> 
> This is not the same as our higher level virtual functions that say get called once per walker per step. Rng being virtual creates is by my estimation 2-5 orders of magnitude more vtable hits per walker per move.
> 
> Changing the RandomBase API a bit for makeGaussRandomWithEngine can mitigate the virtual overhead there by getting blocks of random number to use in assigns without hitting the vtable every single call.
> 

I created a small benchmark to see how much overhead runtime polymorphism adds. On my laptop it's around 10%.

```
$ ./g_benchmark 
2023-06-30T11:31:24-04:00
Running ./g_benchmark
Run on (16 X 3498.04 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 1280 KiB (x8)
  L3 Unified 24576 KiB (x1)
Load Average: 2.24, 2.12, 1.42
-------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations
-------------------------------------------------------------------
VirtualFixture/StdRandom       5.72 ns         5.72 ns    123571260
RawFixture/StdMt19937          5.21 ns         5.21 ns    133855574
```

**Comment by quantumsteve** [comment]
@PDoakORNL Here's the benchmark code
https://github.com/QMCPACK/random_base

**Comment by ye-luo** [comment]
> > I'm a little concerned about functions such as NonLocalECPotential::mw_evaluateImpl where the single rng() calls really really stack up. Unless the compiler or hardware does something clever this looks like a very large number of calls that now have the overhead of being virtual.
> > This is not the same as our higher level virtual functions that say get called once per walker per step. Rng being virtual creates is by my estimation 2-5 orders of magnitude more vtable hits per walker per move.
> > Changing the RandomBase API a bit for makeGaussRandomWithEngine can mitigate the virtual overhead there by getting blocks of random number to use in assigns without hitting the vtable every single call.
> 
> I created a small benchmark to see how much overhead runtime polymorphism adds. On my laptop it's around 10%.
> 
> ```
> $ ./g_benchmark 
> 2023-06-30T11:31:24-04:00
> Running ./g_benchmark
> Run on (16 X 3498.04 MHz CPU s)
> CPU Caches:
>   L1 Data 48 KiB (x8)
>   L1 Instruction 32 KiB (x8)
>   L2 Unified 1280 KiB (x8)
>   L3 Unified 24576 KiB (x1)
> Load Average: 2.24, 2.12, 1.42
> -------------------------------------------------------------------
> Benchmark                         Time             CPU   Iterations
> -------------------------------------------------------------------
> VirtualFixture/StdRandom       5.72 ns         5.72 ns    123571260
> RawFixture/StdMt19937          5.21 ns         5.21 ns    133855574
> ```

Thank you @quantumsteve. Consider the overall time spent on RNG is about 1%, we can afford an additional burden of 10%.

**Review by ye-luo — APPROVED** [review]

**Comment by PDoakORNL** [comment]
I did some more benchmarking in more "realistic" settings i.e. closer to the call profile in the actual code the results are actually even more in favor of the virtual call being close to 0 cost.  In release the timings are +/- %1 for releastic calls:

```
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
benchmark_random is a Catch v2.13.9 host application.
Run with -? for options

-------------------------------------------------------------------------------
makeGaussRandomWithEngine TinyVec bsize 16 1thread StdRandomNoBase
-------------------------------------------------------------------------------
/Users/Shared/ornldev/code/qmcpack/src/Utilities/tests/benchmark_Random.cpp:59
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Std Batched nw:64 stp:128 ps:64                100             1     3.14914 s 
                                        31.7901 ms    31.5706 ms    32.0754 ms 
                                        1.26297 ms    1.02827 ms    1.54555 ms 
                                                                               
Std Batched nw:64 stp:128 ps:128               100             1     6.49297 s 
                                          63.61 ms    63.2888 ms    63.9796 ms 
                                        1.76187 ms    1.54126 ms    2.03827 ms 
                                                                               
Std Batched nw:64 stp:128 ps:256               100             1     13.0473 s 
                                        127.243 ms    126.795 ms    127.725 ms 
                                        2.36169 ms    2.12812 ms    2.71866 ms 
                                                                               

-------------------------------------------------------------------------------
makeGaussRandomWithEngine TinyVec bsize 16 1thread 2 RandomBase child
-------------------------------------------------------------------------------
/Users/Shared/ornldev/code/qmcpack/src/Utilities/tests/benchmark_Random.cpp:114
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Std Batched nw:64 stp:128 ps:64                100             1     3.31951 s 
                                        32.0752 ms    31.9107 ms    32.2993 ms 
                                         973.91 us    759.873 us    1.23555 ms 
                                                                               
Std Batched nw:64 stp:128 ps:128               100             1     6.35493 s 
                                        63.3931 ms    63.0177 ms    63.8984 ms 
                                        2.20249 ms    1.77098 ms    2.88413 ms 
                                                                               
Std Batched nw:64 stp:128 ps:256               100             1     13.1211 s 
                                        126.339 ms    125.812 ms    127.049 ms 
                                        3.09706 ms    2.43252 ms    4.26202 ms 
                                                                               
Fake Batched nw:64 stp:128 ps:256              100             1    991.072 ms 
                                        9.77123 ms    9.66947 ms    9.98137 ms 
                                        711.939 us    430.573 us    1.29436 ms 
                                                                               
Fake Batched nw:64 stp:128 ps:128              100             1      1.9419 s 
                                          21.03 ms     20.716 ms    21.4536 ms 
                                        1.84625 ms    1.43544 ms    2.59582 ms 
                                                                               
Fake Batched nw:64 stp:128 ps:256              100             1     4.33786 s 
                                        41.3482 ms    41.0017 ms    41.7246 ms 
                                        1.83917 ms    1.64344 ms    2.07801 ms 
                                                                               

-------------------------------------------------------------------------------
Nonlocal single StdRandomNoBase
-------------------------------------------------------------------------------
/Users/Shared/ornldev/code/qmcpack/src/Utilities/tests/benchmark_Random.cpp:241
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Std Nonlocal nw:256 stp:128 ps:256             100             1    60.2182 ms 
                                        614.655 us    598.722 us    651.982 us 
                                        117.123 us    63.1771 us    234.311 us 
                                                                               

-------------------------------------------------------------------------------
Nonlocal single thread 2 child
-------------------------------------------------------------------------------
/Users/Shared/ornldev/code/qmcpack/src/Utilities/tests/benchmark_Random.cpp:270
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
Std Nonlocal nw:256 stp:128 ps:256             100             1    60.0351 ms 
                                        595.554 us    582.236 us     618.45 us 
                                        86.8667 us    57.7472 us    129.128 us 
                                                                               
Fake Nonlocal nw:256 stp:128 ps:256            100             1    62.0314 ms 
                                        583.576 us    576.381 us     594.76 us 
                                        45.1936 us     31.978 us    61.3277 us 
                                                                               

===============================================================================
test cases: 4 | 4 passed
assertions: - none -
```

**Comment by PDoakORNL** [comment]
This is with clang 16, osx, x86 8 core intel i9

**Pull request merged**